doc/sphinx/proof-engine/ssreflect-proof-language.rst

```diff
@@ -3637,6 +3636,272 @@ rewriting rule is stated using Leibniz equality (as opposed to setoid
+:ref:`one-liner mode <under_one_liner>` with a ``do`` clause,
+which are explained in more detail below.
+
+.. _under_interactive:
+
+Interactive mode
+````````````````
+
+Let us redo the running example in interactive mode.
+
+.. example::
+
+   .. coqtop:: all abort
+
+      Lemma example_map l : sumlist (map (fun m => m - m) l) = 0.
+      under eq_map => m.
+        rewrite subnn.
+        by over.
+
+The execution of the Ltac expression:
+
+:n:`under @term => [ @i_item__1 | … | @i_item__n ].`
+
+involves the following steps:
+
+1. It performs a :n:`rewrite @term`
+   without failing like in the first example with ``rewrite eq_map.``,
+   but creating evars (see :tacn:`evar`). If :n:`term` is prefixed by
+   a pattern or an occurrence selector, then the modifiers are honoured.
+
+2. As a n-branches intro pattern is provided :tacn:`under` checks that
+   n+1 subgoals have been created. The last one is the main subgoal,
+   while the other ones correspond to premises of the rewrite rule (such as
+   ``forall n, F1 n = F2 n`` for ``eq_map``).
+
+3. If so :tacn:`under` puts these n goals in head normal form (using
+   the defective form of the tactic :tacn:`move`), then executes
+   the corresponding intro pattern :n:`@ipat__i` in each goal.
+
+4. Then :tacn:`under` checks that the first n subgoals
+   are (quantified) equalities or double implications between a
+   term and an evar (eg ``m - m = ?F2`` in the running example).
+
+5. If so :tacn:`under` protects these n goals against an
+   accidental instantiation of the evar.
+   These protected goals are displayed using the ``Under[ … ]``
+   notation (eg ``Under[ m - m ]`` in the running example).
+
+6. The expression inside the ``Under[ … ]`` notation can be
+   proved equivalent to the desired expression
+   by using a regular :tacn:`rewrite` tactic.
+
+7. Interactive editing of the first n goals has to be signalled by
+   using the :tacn:`over` tactic or rewrite rule (see below).
+
+8. Finally, a post-processing step is performed in the main goal
+   to keep the name(s) for the bound variables chosen by the user in
+   the intro pattern for the first branch.
+
+.. _over_ssr:
+
+The over tactic
++++++++++++++++
+
+Two equivalent facilities (a terminator and a lemma) are provided to
+close intermediate subgoals generated by :tacn:`under` (i.e. goals
+displayed as ``Under[ … ]``):
+
+.. tacn:: over
+   :name: over
+
+   This terminator tactic allows one to close goals of the form
+   ``'Under[ … ]``.
+
+.. tacv:: by rewrite over
+
+   This is a variant of :tacn:`over` in order to close ``Under[ … ]``
+   goals, relying on the ``over`` rewrite rule.
+
+.. _under_one_liner:
+
+One-liner mode
+``````````````
+
+The Ltac expression:
+
+:n:`under @term => [ @i_item__1 | … | @i_item__n ] do [ @tac__1 | … | @tac__n ].`
+
+can be seen as a shorter form for the following expression:
+
+:n:`under @term => [ @i_item__1 | … | @i_item__n | ]; [ @tac__1; over | … | @tac__n; over | cbv beta iota ].`
```

Comment:
@CohenCyril @gares 
I've just realized this line could be improved to perfectly describe the semantics of the Ltac expr on previous [Line 3805](https://github.com/coq/coq/pull/9651/files#diff-d80ca9da497679f1161bfbf7e7ab7ce1R3805):

```suggestion
 :n:`under @term; [ move ..| ]; move => [ @i_item__1 | … | @i_item__n | ]; [ @tac__1; over | … | @tac__n; over | cbv beta iota ].`
```

but  I propose to commit this only later on (after Cyril's feedback on the doc itself)

Reply:
I've mixed feeling s about that. What you write is correct, but it also makes the example harder to read and its message more complex: before it was about the extra branch, now it is also about the automatic `move`.
While the latter is useful in practice (when the ipat is omitted) it is just a detail (already documented), while the number of branches not matching the number of goals but still be accepted is way more surprising to me, and  IMO should be the main focus of the example.

BTW, is the extra move always performed or just when the ipat is omitted (and the default `=>*` added)? I think the call to `move` (to do an `hnf`) should only occur in the case the ipat is omitted.

Reply:
Hi @gares, thanks for your comment. I fully agree that the example was harder to read and the message was thereby more difficult to grasp. So I just did another push, replacing `under @term; [ move ..| ] => ...` with `(under @term) => ...` because without the parentheses, the initial example was wrong regarding the specification of the extra branch.

> BTW, is the extra move always performed or just when the ipat is omitted (and the default `=>*` added)? I think the call to `move` (to do an `hnf`) should only occur in the case the ipat is omitted.

Good question: actually, the extra (hnf-like) `move.` is always performed, and it has two interesting benefits:

1. it improve user-experience with lemmas such as `eq_big` (e.g. `under eq_big=>[*|i Hi] do [rewrite something|...]`) because if the `move.` was not performed in this case, due to the automatic protection with `'Under[ _ ]` on the conclusion of the goal, **some kind of hnf would basically be performed on the goal** (compare `erewrite eq_big.` with `under eq_big.`) but without the `; [move ..|]` the outcome of `=> [*|...]` would sound weird to the user...

2. thanks to this `move.` which is always done at the proper time (and already documented in the `_under_interactive` section as you pointed out), the final example I pushed is formally correct:
```
(under @term) => [ @i_item__1 | … | @i_item__n | ]; [ @tac__1; over | … | @tac__n; over | cbv beta iota ].
```